src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -382,289 +428,317 @@ private static HttpRequestException GetVersionException(HttpRequestMessage reque
             return new HttpRequestException(SR.Format(SR.net_http_requested_version_cannot_establish, request.Version, request.VersionPolicy, desiredVersion));
         }
 
-        private ValueTask<HttpConnection?> GetOrReserveHttp11ConnectionAsync(bool async, CancellationToken cancellationToken)
+        private bool CheckExpirationOnGet(HttpConnectionBase connection)
+        {
+            TimeSpan pooledConnectionLifetime = _poolManager.Settings._pooledConnectionLifetime;
+            if (pooledConnectionLifetime != Timeout.InfiniteTimeSpan)
+            {
+                return connection.GetLifetimeTicks(Environment.TickCount64) > pooledConnectionLifetime.TotalMilliseconds;
+            }
+
+            return false;
+        }
+
+        private async Task AddHttp11Connection(HttpRequestMessage request)
+        {
+            if (NetEventSource.Log.IsEnabled()) Trace("Creating new HTTP/1.1 connection for pool.");
+
+            HttpConnection connection;
+            try
+            {
+                using (CancellationTokenSource cts = GetConnectTimeoutCancellationTokenSource())
+                {
+                    connection = await CreateHttp11ConnectionAsync(request, false, cts.Token).ConfigureAwait(false);
+                }
+            }
+            catch (Exception e)
+            {
+                HandleHttp11ConnectionFailure(e);
+                return;
+            }
+
+            // Add the established connection to the pool.
+            ReturnHttp11Connection(connection, isNewConnection: true);
+        }
+
+        private void CheckForHttp11ConnectionInjection()
         {
-            long nowTicks = Environment.TickCount64;
+            Debug.Assert(Monitor.IsEntered(SyncObj));
+
+            if (!_http11RequestQueue.TryPeekNextRequest(out HttpRequestMessage? request))
+            {
+                return;
+            }
 
+            // Determine if we can and should add a new connection to the pool.
+            if (_availableHttp11Connections.Count == 0 &&                           // No available connections
+                _http11RequestQueue.Count > _pendingHttp11ConnectionCount &&        // More requests queued than pending connections
+                _associatedHttp11ConnectionCount < _maxHttp11Connections)           // Under the connection limit
+            {
+                _associatedHttp11ConnectionCount++;
+                _pendingHttp11ConnectionCount++;
+
+                Task.Run(() => AddHttp11Connection(request));
```

Comment:
Why do we need to queue a work item to call this?

Reply:
First, because we are holding the lock here. We could move the AddHttp11Connection outside the lock with a bit of work if we wanted. 

Second, it seems good in general to queue any work associated with creating the new connection vs starting to do that work on the calling thread. This routine is called in a couple different places, and if we were to call AddHttp11Connection directly here, we might be delaying some other work from completing. 

Thoughts?